### ISSUE TITLE
Create a script API method to be executed on all workers finish

### ISSUE THREAD

**Issue by patrickdalla** [sepinf-inc/IPED#1554]
The task class is instantiated once per worker. The current finish method executes after the current Worker where the task instance is allocated is finished. So it is executed for every worker. 
It is enough for instance resource release. But for some tasks that should be executed in the case end scope, only once per case, it gets repeated per worker thread.
For example, in query result bookmark creation script, it get executed for every worker end. At the end, in this specific case, the result is the same thought. But for other use cases, it can be a problem.
In scripts I've done, I could achieve the script relevant code only once making a logic to count each instance init and later decreasing this counter until it reaches the last worker being finished. But it may be not intuitive to the end user.

At first I thought it could be a method of the class AbstractTask. But it can still be a little confusing. What do you think @lfcnassif ?

**Comment by patrickdalla** [comment]
I think that new method in AbstractTask can gets confusing because it is designed to be executed per item.
But creating another point to include scripts, following another interface rule, can lead to excessive code maintenance.

**Comment by lfcnassif** [comment]
@bhoelz suggested this years ago... I do the same trick (count the number of inits) in one place, in others I simply call `worker.getManager().getNumWorkers()` but it is decremented in finish() anyway. I completely agree it is not intuitive. IPED already have old features that could benefit of this like the `P2PBookmarker` class and `Manager.filterKeywords()` method, both called in a hard coded way.

A new interface would be cleaner, but would increase maintenance as you said (scripts installation, more python scripts managing code, etc). What about a new `finishAfterAll()` method in AbstractClass, like AfterAll() in unit tests?

**Comment by lfcnassif** [comment]
Just updated the title because this can be extended by java code too.

**Comment by patrickdalla** [comment]
maybe the name could be "executeAfterAll" and it would enqueue.
And about its execution? It should be called on last Worker thread? In this case, should it be called before this last worker finish() method? Or we enqueue it to be called outside Worker thread, always after finish() method, in which case the resources released in finish() wouldn't be accessible?

**Comment by lfcnassif** [comment]
> maybe the name could be "executeAfterAll".

I think I still prefer `finishAfterAll`. Of course the docs should be clear about the difference. Any opinion @tc-wleite?

> And about its execution? It should be called on last Worker thread? In this case, should it be called before this last worker finish() method? Or we enqueue it to be called outside Worker thread, always after finish() method, in which case the resources released in finish() wouldn't be accessible?

I thought to execute after all Worker/Tasks threads are finished, this can bring some benefit, in Manager main thread. But maybe PythonTask wouldn't work, if I remember they should run on the same thread from which the python interpreter was created... Do you need the Task resources? The idea to create another interface/point of extension wouldn't access Task resources too, I think. What is your use case?

**Comment by patrickdalla** [comment]
In one of my use case I was linking $I trash files with their correspondent $R file to export in the linked info liked full path, name and deletion date (found in $I) and the real content, access modified record and creation dates (found in $R), exporting this data to a html formatted report.
At time of $I item process, maybe there wasn't the respective $R processed with its metadata on case yet, and vice-versa. So it needed to be executed on the end.
I don't think I need the task resources, at least in this use case. 
I was really asking to know your opinion/experience about requirements, like this supposed requirement of Python interpreter executing on the same creation thread. Is it true?
If so, the solution could be implement the same "trick" of worker counting, but in java, and call finishAfterAll on the after last worker finish();

**Comment by patrickdalla** [comment]
As for the method name, finish seems something to be done to gracefully END the task, and seems a good choice for the already existent method, that makes part of the task instance lifecycle.
But the real objective of this new method is to execute a code that will create something (bookmark, reports, metadatas). I think that the term execute more appropriate. 
But I can use finishAfterAll.

**Comment by lfcnassif** [comment]
> I was really asking to know your opinion/experience about requirements, like this supposed requirement of Python interpreter executing on the same creation thread. Is it true?

It's true. But if I remember if PythonTask is called from a new Thread, it should detect it and create a new python interpreter for that thread, so I think calling it from the main Manager thread could work.

**Comment by tc-wleite** [comment]
> I think I still prefer `finishAfterAll`. Of course the docs should be clear about the difference. Any opinion @tc-wleite?

Sorry, I forgot to reply here.
I was following the discussion, but not that closely.
In general, the worker would need to access its resources (including static ones) to do something when the processing is finished. So I guess calling this method before `finish()` may be useful in some cases . But in others, it can make sense to call after the last (e.g. data needed to be saved/commited or files needed to be close, to do something else).
Currently, with `finish()`, the task implementation has full control and can decide what to do before the first or after the last `finish()` call, right?
What about keeping only `finish()` and adding a utility method in parent class (or a parameter) to allow checking if it is the first and/or the last execution of `finish()`?
That would also be useful in the initialization.
Obviously, that only works if `init()` and `finish()` are called sequentially (not in parallel).

**Comment by tc-wleite** [comment]
> But if I understood correctly, the idea about the new method would be to execute it after all Tasks have completely finished, so, for instance, IndexTask would have committed all data, like you said, and we could search the case data to create Bookmarks, etc, without having to worry about Task pipeline position, this is an advantage.

Oh, I see. I missed that part.
Then it does make sense to have a new method, not sure about the best name though.
I was wrongly thinking that the goal was to avoid the code for controlling if it is the first or last instance to be initiated or terminated in each task implementation.

**Comment by lfcnassif** [comment]
> I was wrongly thinking that the goal was to avoid the code for controlling if it is the first or last instance to be initiated or terminated in each task implementation.

This is annoying for sure, so your utility method is useful!

**Comment by patrickdalla** [comment]
Right, maybe a utility method can solve the issue. It is simpler, and does not alters the interface.
